drivers/audio/pdm/pdm_nrfx.c

```diff
@@ -0,0 +1,229 @@
+	struct k_work work;
+	int16_t *buffer;
+};
+
+struct pdm_nrfx_data {
+	pdm_data_handler_t data_handler;
+	struct k_mem_slab *mem_slab;
+	struct buffer_released_data buff_released;
+	uint8_t active_buffer;
+};
+
+static int16_t m_next_buffer[PDM_NRFX_BUFFER_SIZE][PDM_NRFX_NUMBER_OF_BUFFERS];
+
+static struct pdm_nrfx_data m_pdm_nrfx_data;
+
+static void release_buffer(struct k_work *item)
+{
+	struct device *dev = DEVICE_GET(pdm_0);
+
+	struct pdm_nrfx_data *driver_data = dev->driver_data;
+
+	struct buffer_released_data *data =
+		CONTAINER_OF(item, struct buffer_released_data, work);
+
+	driver_data->data_handler(data->buffer, PDM_NRFX_BUFFER_SIZE);
+}
+
+static void pdm_nrfx_event_handler(nrfx_pdm_evt_t const *p_evt)
+{
+	struct device *dev = DEVICE_GET(pdm_0);
+	struct pdm_nrfx_data *driver_data = dev->driver_data;
+
+	void *buffer;
+
+	if (p_evt->error == NRFX_PDM_ERROR_OVERFLOW) {
+		LOG_ERR("Overflow error when handling event.\n");
+		return;
+	}
+
+	if (nrfx_pdm_enable_check()) {
+		/**
+		 * If a buffer was requested,
+		 * provide a new one, and alternate the active buffer
+		 */
+		if (p_evt->buffer_requested) {
+			nrfx_err_t result;
+
+			result = nrfx_pdm_buffer_set(
+				 &m_next_buffer[0][driver_data->active_buffer],
+				 PDM_NRFX_BUFFER_SIZE);
+
+			if (result != NRFX_SUCCESS &&
+			    result != NRFX_ERROR_BUSY) {
+				LOG_ERR(
+				"Failed to set new buffer, error %d.\n",
+				result);
+			}
+			driver_data->active_buffer =
+					(~driver_data->active_buffer) & 0x01;
+		}
+
+		/**
+		 * If a buffer has been released,
+		 * save it and submit it to the workqueue for the event handler
+		 */
+		if (p_evt->buffer_released != NULL) {
+			int ret;
+
+			ret = k_mem_slab_alloc(driver_data->mem_slab,
+					       &buffer, K_NO_WAIT);
+			if (ret == -ENOMEM) {
+				LOG_ERR(
+				"Not enough memory to allocate new buffer");
+				return;
+			} else if (ret) {
+				LOG_ERR("Failed to allocate buffer, error: %d",
+					ret);
+				return;
+			}
+			memcpy(buffer, p_evt->buffer_released,
+			       sizeof(int16_t) * PDM_NRFX_BUFFER_SIZE);
+			driver_data->buff_released.buffer = (int16_t *)buffer;
+			k_work_submit(&driver_data->buff_released.work);
+		}
+	}
+}
+
+
+static int pdm_nrfx_configure(struct device *dev, struct pdm_config *cfg)
+{
+	if (!cfg->data_handler || !cfg->mem_slab) {
+		return -EINVAL;
+	}
+	struct pdm_nrfx_data *driver_data = dev->driver_data;
+
+	driver_data->mem_slab = cfg->mem_slab;
+	driver_data->data_handler = cfg->data_handler;
+	return 0;
+}
+
+
+static int pdm_nrfx_start(struct device *dev)
+{
+	struct pdm_nrfx_data *driver_data = dev->driver_data;
+	nrfx_err_t result;
+
+	result = nrfx_pdm_buffer_set(
+			&m_next_buffer[0][driver_data->active_buffer],
+			PDM_NRFX_BUFFER_SIZE);
+
+	if (result != NRFX_SUCCESS) {
+		LOG_ERR("Failed to set new buffer, error %d.\n", result);
+	}
+	driver_data->active_buffer = (~driver_data->active_buffer) & 0x01;
+	result = nrfx_pdm_start();
+	if (result == NRFX_ERROR_BUSY) {
```

Comment:
If the module was busy, it would've indicated that when we did the call to `nrfx_pdm_buffer_set`. So we should move this check after line 151. I also think that we shouldn't proceed forward if we `nrfx_pdm_buffer_set` fails.